language/bytecode_verifier/src/instantiation_loops.rs

```diff
@@ -0,0 +1,202 @@
+    node_map: HashMap<Node, NodeIndex>,
+}
+
+impl<'a> InstantiationLoopChecker<'a> {
+    pub fn new(module: &'a CompiledModule) -> Self {
+        Self {
+            module,
+            graph: Graph::new(),
+            node_map: HashMap::new(),
+        }
+    }
+
+    /// Retrives the node corresponding to the specified type formal.
+    /// If none exists in the graph yet, create one.
+    fn get_or_add_node(&mut self, node: Node) -> NodeIndex {
+        match self.node_map.entry(node) {
+            hash_map::Entry::Occupied(entry) => *entry.get(),
+            hash_map::Entry::Vacant(entry) => {
+                let idx = self.graph.add_node(node);
+                entry.insert(idx);
+                idx
+            }
+        }
+    }
+
+    /// Helper function that extracts type parameters from a given type.
+    /// Duplicated entries are removed.
+    fn extract_type_parameters(ty: &SignatureToken) -> Vec<TypeParameterIndex> {
+        use SignatureToken::*;
+
+        fn rec(ty: &SignatureToken) -> Vec<TypeParameterIndex> {
+            match ty {
+                Bool | Address | U64 | String | ByteArray => vec![],
+                TypeParameter(idx) => vec![*idx],
+                Reference(ty) | MutableReference(ty) => rec(ty),
+                Struct(_, tys) => tys.iter().map(rec).flatten().collect(),
+            }
+        }
+
+        let mut type_params = rec(ty);
+        type_params.sort();
+        type_params.dedup();
+        type_params
+    }
+
+    /// Builds a graph G such that
+    ///   - Each type formal of a generic function is a node in G.
+    ///   - There is an edge from type formal f_T to g_T if f_T is used to instantiate g_T in a
+    ///     call.
+    ///     - Each edge is labeled either `Identity` or `TyConApp`. See `Edge` for details.
+    fn build_graph(&mut self) {
+        for def in self
+            .module
+            .function_defs()
+            .iter()
+            .filter(|def| !def.is_native())
+        {
+            for instr in &def.code.code {
+                if let Bytecode::Call(other_func_idx, type_actuals_idx) = instr {
+                    let other_func_handle = self.module.function_handle_at(*other_func_idx);
+                    if other_func_handle.module != ModuleHandleIndex::new(0) {
+                        continue;
+                    }
+
+                    let type_actuals = &self.module.locals_signature_at(*type_actuals_idx).0;
+                    for (formal_idx, ty) in type_actuals.iter().enumerate() {
+                        let formal_idx = formal_idx as TypeParameterIndex;
+                        let node_to = self.get_or_add_node(Node(*other_func_idx, formal_idx));
+                        if let SignatureToken::TypeParameter(actual_idx) = ty {
+                            let node_from = self.get_or_add_node(Node(def.function, *actual_idx));
+                            self.graph.add_edge(node_from, node_to, Edge::Identity);
```

Comment:
Is this an `Identity` here? Even if `actual_idx != formal_idx`? If so, maybe rename `Identity`?

Reply:
The identity transformation is not about the type variables themselves, but the actual (opaque) types bound to them.
```
bar<T1, T2>() {
}

foo<T>() {
    bar<u64, T>()
}
```
Here when we're calling `bar`, conceptually we
1) look up the type bound to type variable T in the scope of foo
2) bind the type to type variable T2 in the scope of bar.

The type, what ever it is, is unchanged during this process. This is why I'm calling this type of edge "Identity".